src/lib/UXDController.ts

```diff
@@ -28,158 +31,181 @@ export class UXDController {
-  public mintWithEth({
-    market,
-    ethAmount,
+  public async mint({
+    amount,
     slippage,
     signer,
+    collateralToken,
   }: {
-    market: string;
-    ethAmount: BigNumber;
-    slippage: BigNumber;
+    amount: number;
+    slippage: number;
     signer: Signer;
+    collateralToken?: string;
   }): Promise<ContractTransaction> {
-    return this.controllerContract.connect(signer).mintWithEth(
-      market,
-      slippage,
-      {
-        value: ethAmount,
-      },
-    );
-  }
-
-  public redeemEth({
-    market,
-    uxdAmount,
+    const ethAmount = ethers.utils.parseEther(amount.toString());
+    const ethSlippage = ethers.utils.parseEther(slippage.toString());
+    if (arguments.length == 4) {
+      return await this.mintWithERC20(
+        ethAmount,
+        ethSlippage,
+        signer,
+        collateralToken,
+      );
+    } else if (arguments.length == 3) {
+      return await this.mintWithETH(ethAmount, ethSlippage, signer);
+    } else {
+      throw Error("Invalid number of arguments");
+    }
+  }
+
+  private async mintWithERC20(
+    ethAmount: BigNumber,
+    ethSlippage: BigNumber,
+    signer: Signer,
+    collateral: string,
+  ): Promise<ContractTransaction> {
+    return await this.controllerContract
+      .connect(signer)
+      .mint(this.market, collateral, ethAmount, ethSlippage);
+  }
+
+  private async mintWithETH(
+    ethAmount: BigNumber,
+    ethSlippage: BigNumber,
+    signer: Signer
+  ): Promise<ContractTransaction> {
+    return await this.controllerContract
+      .connect(signer)
+      .mintWithEth(this.market, ethSlippage, { value: ethAmount });
+  }
+
+  public async redeem({
+    amount,
     slippage,
     signer,
+    collateralToken,
   }: {
-    market: string;
-    uxdAmount: BigNumber;
-    slippage: BigNumber;
+    amount: number;
+    slippage: number;
     signer: Signer;
+    collateralToken: string;
   }): Promise<ContractTransaction> {
-    return this.controllerContract.connect(signer).redeemEth(market, uxdAmount, slippage);
+    const uxdAmount = ethers.utils.parseEther(amount.toString());
+    const slippageAmount = ethers.utils.parseEther(slippage.toString());
+    if (arguments.length == 4) {
+      return await this.redeemERC20(uxdAmount, slippageAmount, signer, collateralToken)
+    } else if (arguments.length == 3) {
+      return await this.redeemEth(uxdAmount, slippageAmount, signer);
+    } else {
+      throw Error("Invalid number of arguments");
+    }
+  }
+
+  private async redeemERC20(
+    uxdAmount: BigNumber,
+    slippageAmount: BigNumber,
+    signer: Signer,
+    collateral: string,
+  ): Promise<ContractTransaction> {
+    return await this.controllerContract
+      .connect(signer)
+      .redeem(this.market, collateral, uxdAmount, slippageAmount);
+  }
+  private async redeemEth(
+    uxdAmount: BigNumber,
+    slippageAmount: BigNumber,
+    signer: Signer
+  ): Promise<ContractTransaction> {
+    return await this.controllerContract
```

Comment:
Is there a reason for this `await` to be here?

Reply:
Removed